### PR TITLE
MEN-3131: integration test for the mtls-ambassador and mtls-enabled client

### PR DESCRIPTION
**Pull request by tranchitella**
Changelog: none

Signed-off-by: Fabio Tranchitella <fabio.tranchitella@northern.tech>

### CODE REVIEW

File: tests/tests/test_mtls.py [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+        tenant = setup_ent_mtls.tenant
+
+        # stop the api gateway
+        setup_ent_mtls.stop_api_gateway()
+
+        # start a new mender client
+        setup_ent_mtls.new_mtls_client("mender-client", tenant.tenant_token)
+
+        # stop the Mender client
+        device = MenderDevice(setup_ent_mtls.get_mender_clients()[0])
```
Comment (lluiscampos) [inline tests/tests/test_mtls.py:103]: For the test framework to find the `MenderDevice`, the object needs to be referenced from the setup:

```suggestion
        setup_ent_mtls.device = MenderDevice(setup_ent_mtls.get_mender_clients()[0])
```

Comment (tranchitella) [reply tests/tests/test_mtls.py]: Done!

File: tests/tests/test_mtls.py [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+                "extra",
+                "mtls",
+                "certs",
+                "client",
+            ),
+            remote_path="/etc/mender",
+        )
+        device.run("update-ca-certificates")
+
+        # TODO: this is needed to circumnavigate a bug in mender-client, which
```
Comment (oleorhagen) [inline tests/tests/test_mtls.py:199]: Comment:
This can safely be removed now that we're on the newest sha :)

File: docker-compose.override.yml [hunk 3]
```diff
@@ -0,0 +1,34 @@
+version: '2.1'
```
Comment (lluiscampos) [inline docker-compose.override.yml:1]: File committed by mistake? Please remove

File: tests/tests/test_mtls.py (hunk 2 above)
Comment (tranchitella) [reply tests/tests/test_mtls.py]: Removed

File: tests/tests/test_mtls.py [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+        device.put(
+            "cert.crt",
+            local_path=os.path.join(
+                os.path.dirname(__file__), "..", "..", "certs", "storage-proxy",
+            ),
+            remote_path="/etc/mender",
+        )
+        device.run(
+            "cat /etc/mender/server.crt /etc/mender/cert.crt > /tmp/server.crt && mv /tmp/server.crt /etc/mender/server.crt && rm /etc/mender/cert.crt"
+        )
```
Comment (lluiscampos) [inline tests/tests/test_mtls.py:302]: Up until here the code is the same as the previous text; consider refactoring.

**Review by lluiscampos — COMMENTED** [review]
The tests themselves look good. There is room for refactoring, but being a feature that is unlikely to be extended, I have no strong opinion in leaving them explicit as they are.

 The certs I cannot review...

File: tests/tests/test_mtls.py [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+        try:
+            # retrieve the original configuration file
+            output = device.run("cat /etc/mender/mender.conf")
+            config = json.loads(output)
+            # replace mender.conf with an mTLS enabled one
+            config["ServerURL"] = "https://mtls-ambassador:8080"
+            config["ServerCertificate"] = "/etc/mender/server.crt"
+            config["SkipVerify"] = True
+            config["HttpsClient"] = {
+                "Certificate": "/etc/mender/client.1.%s.crt" % algorithm,
```
Comment (oleorhagen) [inline tests/tests/test_mtls.py:118]: Question:
Why not use `f-strings` here also?

File: tests/tests/test_mtls.py [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+sleep 6s
+
+# Successful update
+exit 0
+"""
+
+    with tempfile.NamedTemporaryFile(suffix="testdeployment") as tf:
+        tf.write(script)
+        tf.seek(0)
+        out = tf.read()
```
Comment (oleorhagen) [inline tests/tests/test_mtls.py:79]: Observation:
Kinda weird there isn't a utility function for this :thinking:

File: tests/tests/test_mtls.py [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+
+    auth.username = mtls_username
+    auth.password = mtls_password
+    auth.multitenancy = True
+    auth.current_tenant = env.tenant
+
+    return env
+
+
+def make_script_artifact(artifact_name, device_type, output_path):
```
Comment (oleorhagen) [inline tests/tests/test_mtls.py:63]: Observation:
It seems this exact same functionality is present in `test_deployments_retry.py`. Maybe we should unify it now whilst we have the chance?

File: tests/tests/test_mtls.py [hunk 8]
```diff
@@ -0,0 +1,315 @@
...
+            # replace mender.conf with an mTLS enabled one
+            config["ServerURL"] = "https://mtls-ambassador:8080"
+            config["ServerCertificate"] = "/etc/mender/server.crt"
+            config["SkipVerify"] = True
+            config["HttpsClient"] = {
+                "Certificate": "/etc/mender/client.1.%s.crt" % algorithm,
+                "Key": "/etc/mender/client.1.%s.key" % algorithm,
+            }
+            del config["ArtifactVerifyKey"]
+            device.run("rm /data/mender/mender.conf")
```
Comment (oleorhagen) [inline tests/tests/test_mtls.py:122]: Question:
Why is this removal needed?

File: tests/tests/test_mtls.py [hunk 9]
```diff
@@ -0,0 +1,315 @@
...
+                "certs",
+                "tenant-ca",
+            ),
+            remote_path="/etc/ssl/certs",
+        )
+        device.put(
+            "server.crt",
+            local_path=os.path.join(
+                os.path.dirname(__file__),
+                "..",
```
Comment (oleorhagen) [inline tests/tests/test_mtls.py:152]: Question:
Is this not a bit excessive? We're not testing on anything but Unix. So the separator is always going to be `/`. Is not hardcoding this a bit easier to read?

Comment (oleorhagen) [reply tests/tests/test_mtls.py:152]: Also, the `../../extra/mtls/certs` prefix seems to be used a lot of places.

**Review by oleorhagen — COMMENTED** [review]
Looks good. 

Just left some questions and comments, feel free to ignore at will

File: docker-compose.override.yml (hunk 3 above)
Comment (tranchitella) [reply docker-compose.override.yml:1]: sorry, you are right

File: tests/tests/test_mtls.py (hunk 9 above)
Comment (tranchitella) [reply tests/tests/test_mtls.py:152]: I prefer to stick to os.path.join, but I refactored it a bit to use base path variables.

File: tests/tests/test_mtls.py (hunk 8 above)
Comment (tranchitella) [reply tests/tests/test_mtls.py:122]: not needed, removed

File: tests/tests/test_mtls.py (hunk 6 above)
Comment (tranchitella) [reply tests/tests/test_mtls.py:79]: I am not aware of an utility to write and read back from the file with a single instruction...

File: tests/tests/test_mtls.py (hunk 4 above)
Comment (tranchitella) [reply tests/tests/test_mtls.py:302]: refactored

File: tests/tests/test_mtls.py (hunk 5 above)
Comment (tranchitella) [reply tests/tests/test_mtls.py:118]: let's use f-strings :)

File: tests/tests/test_mtls.py (hunk 7 above)
Comment (tranchitella) [reply tests/tests/test_mtls.py:63]: Refactored

**Comment by tranchitella** [comment]
@lluiscampos @oleorhagen thank you for your feedback, I reworked this PR to apply the changes you suggested.
Tests pass locally (with a minor glitch that needs to be fixed in https://github.com/mendersoftware/mender/pull/588, already explained to @oleorhagen).

Can you please have a look at it again?

File: tests/tests/test_mtls.py (hunk 6 above)
Comment (oleorhagen) [reply tests/tests/test_mtls.py]: Shame, `tempfile.WithData("foobar")` would have been nice

**Review by oleorhagen — APPROVED** [review]
Looks good imo :)

**Comment by tranchitella** [comment]
New pipeline: https://gitlab.com/Northern.tech/Mender/mender-qa/-/pipelines/181632818

**Review by lluiscampos — APPROVED** [review]
Thank you for refactoring. Looks good!

**Comment by tranchitella** [comment]
Yet another pipeline: https://gitlab.com/Northern.tech/Mender/mender-qa/-/pipelines/181870265

**Comment by lluiscampos** [comment]
@tranchitella The last force-push looks good to me. Sorry I didn't catch up before the problem with the explicit "master" version in the client docker-compose file.

**Comment by tranchitella** [comment]
@lluiscampos you saved our day identifying the issue, it was me who made it wrong :)

**Review by mchalski — APPROVED** [review]
